### ISSUE TITLE
Amesos2 doesn't compile with Tacho disabled

### ISSUE THREAD

**Issue by ibaned** [trilinos/Trilinos#1772]
I just got a compile error while running the checkin script for #1767:

```
In file included from /scratch/daibane/Trilinos/packages/amesos2/src/Amesos2_Tacho.hpp:47:0,
                 from /scratch/daibane/Trilinos/packages/amesos2/src/Amesos2_Factory.hpp:111,
                 from /scratch/daibane/Trilinos/packages/amesos2/src/Amesos2_Details_LinearSolverFactory_def.hpp:52,
                 from /scratch/daibane/Trilinos/packages/amesos2/src/Amesos2_Details_LinearSolverFactory.cpp:51:
/scratch/daibane/Trilinos/packages/amesos2/src/Amesos2_Tacho_decl.hpp:51:21: fatal error: Tacho.hpp: No such file or directory
 #include "Tacho.hpp"
                     ^
```

@trilinos/amesos2  @kyungjoo-kim  this is breaking all attempts to use the checkin script.
Please make sure you use the checkin script before pushing changes, or at least test that Amesos2 still compiles when ShyLUTacho is disabled.

**Comment by kyungjoo-kim** [comment]
Sorry. I merged the PR. I will fix this.

**Comment by ibaned** [comment]
I guess the PR in question is #1563

**Comment by kyungjoo-kim** [comment]
Yes correct. I just need to change the status of my code to "ST" from "EX". I am running my check in script to verify this. It won't take much of time (probably 30 mins).

**Comment by bartlettroscoe** [comment]
Here is the push that broke the standard CI build:

```
Mon Sep 25 14:35:10 MDT 2017

commit 5ed34f8377576a8cc1310d8f39e5956cdfa60023
Merge: 6240ccb d044af2
Author:     kyungjoo-kim <kyukim@sandia.gov>
AuthorDate: Mon Sep 25 14:34:54 2017 -0600
Commit:     GitHub <noreply@github.com>
CommitDate: Mon Sep 25 14:34:54 2017 -0600

    Merge pull request #1563 from Tech-XCorp/amesos2Tacho
    
    Amesos2: Tacho Wrapper Setup

Commits pushed:
5ed34f8 Merge pull request #1563 from Tech-XCorp/amesos2Tacho
6240ccb EpetraExt: fix memspace handle closing
d044af2 Amesos2: Tacho ideas for optimization
6dba603 Amesos2: Tacho use multiple rhs
f3db6c7 Amesos2: Tacho remove unused typedefs
70b174a Amesos2: Tacho remove specialization comments
4b8d82b Amesos2: Tacho make workspace allocated once.
ae97149 Amesos2: Tacho remove unused members.
aed2636 Amesos2: Tacho disable single process optimization
06e7079 Amesos2: Tacho remove complex check
8c3bbd4 Tacho: Update test cmake dependencies
e3ba67e Amesos2: Tacho Wrapper Update (#1563)
7fe74b8 Amesos2: Tacho Wrapper Setup
```

(FYI: Looks like @ibaned's commit `6240ccb "EpetraExt: fix memspace handle closing"` just happened to get pushed at the same time, which is why it is shown above.  This is the weakness of the push logger I wrote.  It can't resolve pushes that occur at about exactly the same time.  In this case, they were pushed within 1 minute, which is very rare.)

This build failure is now shown in the post-push CI build in the latest iteration:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3127373

I don't see any commits pushed after that that would fix this.

It would be good to fix this before the nighty builds fire off (or back out the commits that broke this).

If no fixing commit is pushed in the next hour, then I will revert the merge commit:

```
5ed34f8 "Merge pull request #1563 from Tech-XCorp/amesos2Tacho"
Author: kyungjoo-kim <kyukim@sandia.gov>
Date:   Mon Sep 25 14:34:54 2017 -0600 (4 hours ago)
```

and push (perhaps with not much testing or just testing Amesos2 since @ibaned's push used the checkin-test-sems.sh script so that commit should be fine).

**Comment by bartlettroscoe** [comment]
I reverted this with:

```
$ cd Trilinos/
$ git revert -m 1 5ed34f8377576a8cc1310d8f39e5956cdfa60023
```

This is now identical to the version of Trilinos in commit:

```
commit 6240ccb6371fe0243648561a2f258b4ce2f21c75
Author: Dan Ibanez <daibane@sandia.gov>
Date:   Sun Sep 24 17:27:31 2017 -0600

    EpetraExt: fix memspace handle closing

M       packages/epetraext/src/inout/EpetraExt_HDF5.cpp
```

as shown by:

```
$ git diff HEAD 6240ccb6371fe0243648561a2f258b4ce2f21c75
[empty]
```

I am running:

```
$ ./checkin-test-sems.sh --no-enable-fwd-packages --do-all --push
```

This will test the modified packages only (ShyLU and Amesos2) and then push.  Should be much better than leaving a build failure to go into nightly testing.

**Comment by bartlettroscoe** [comment]
Last night I pushed the revert commit:

```
Author:     Roscoe A. Bartlett <rabartl@sandia.gov>
AuthorDate: Mon Sep 25 20:06:55 2017 -0600
Commit:     Roscoe A. Bartlett <rabartl@sandia.gov>
CommitDate: Mon Sep 25 20:18:04 2017 -0600

    Revert "Merge pull request #1563 from Tech-XCorp/amesos2Tacho"
    
    This reverts commit 5ed34f8377576a8cc1310d8f39e5956cdfa60023, reversing
    changes made to 6240ccb6371fe0243648561a2f258b4ce2f21c75.
    
    This is needed to fix the standard CI build (see #1772).
    
    Build/Test Cases Summary
    Enabled Packages: Amesos2, ShyLUTacho
    Disabled Packages: PyTrilinos,Claps,TriKota
    0) MPI_RELEASE_DEBUG_SHARED_PT => passed: passed=8,notpassed=0 (3.81 min)

  84.0% packages/amesos2/src/
   4.7% packages/amesos2/test/matrices/
   9.6% packages/amesos2/test/solvers/
```

So far the CI build this morning is looking clean, including for Amesos2 as shown at:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3128858

@kyungjoo-kim,

Please fix this on the branch, then test with:

```
$ cd CHECKIN/
$ ./checkin-test-sems.sh --enable-packages=Amesos2 --local-do-all
```

That will tell you if it is okay to merge this branch.

Or, if you want to just cross your fingers and merge and push, then please do:

```
$ cd Trilinos/
$ git checkout develop
$ git pull
$ git merge Tech-XCorp/amesos2Tacho  # Assuming this is the fixed branch
$ cd CHECKIN/
$ ./checkin-test-sems.sh --do-all --push --no-rebase  # No rebase is important to preserve merge commits
```

Or, if you want to just push from your local branch, you can use the [remote pull/test/push](https://github.com/trilinos/Trilinos/wiki/Local-development-with-remote-pull,-test,-and-push) setup.

Please let me know if you have questions about how to properly merge, test, and push this with the checkin-test-sems.sh script.

**Comment by kyungjoo-kim** [comment]
@bartlettroscoe 

How can I use a kind of autofill of my password when it is pushed. I ran the check-in script yesterday and it took a lot more time than I expected. It is finished after 6 PM and I had to be out of office before 6 PM due to foreign national policy in this building. Is there anyway that I can ran the script with password together ? So after all tests are done, it is pushed without a prompt to ask my password ?

**Comment by ibaned** [comment]
use the SSH protocol when cloning Trilinos (`git@github.com:trilinos/Trilinos.git`), and set up your SSH keys on that machine.

**Comment by bartlettroscoe** [comment]
>>  Is there anyway that I can ran the script with password together ? 
> 
> use the SSH protocol when cloning Trilinos (git@github.com:trilinos/Trilinos.git), and set up your SSH keys on that machine.

Yes, that is correct.  Note the WARNING near the top of the page:

* https://github.com/trilinos/Trilinos/wiki/Policies-%7C-Safe-Checkin-Testing

which says "(**WARNING:** The Trilinos git repo must use SSH with the remote URL git@github.com:trilinos/Trilinos.git and not HTTPS for access to github. See below note)."

SSH is our friend.

**Comment by bartlettroscoe** [comment]
Looks like the fix got pushed today as shown in the push logger:

```
Tue Sep 26 08:01:21 MDT 2017

commit c85894418edee95cd58ae5f9c63f9e5ac4ea68dd
Author:     Kyungjoo Kim <kyukim@sandia.gov>
AuthorDate: Tue Sep 26 01:06:48 2017 -0600
Commit:     Kyungjoo Kim <kyukim@sandia.gov>
CommitDate: Tue Sep 26 07:59:59 2017 -0600

    Amesos2 - fix for #1772
    
    Build/Test Cases Summary
    Enabled Packages: Amesos2
    Disabled Packages: PyTrilinos,Claps,TriKota
    Enabled all Forward Packages
    0) MPI_RELEASE_DEBUG_SHARED_PT => passed: passed=772,notpassed=0 (16.55 min)
    Other local commits for this build/test group: a428ba4

Commits pushed:
c858944 Amesos2 - fix for #1772
a428ba4 Revert "Revert "Merge pull request #1563 from Tech-XCorp/amesos2Tacho""
```

And it got pulled in the CI iteration that started at 15:43 UTC today shown in:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3129197

where these commits are shown pulled in this CI iteration in:

* https://testing.sandia.gov/cdash/viewNotes.php?buildid=3129199##note1

showing:

```
c858944:  Amesos2 - fix for #1772
Author: Kyungjoo Kim <kyukim@sandia.gov>
Date:   Tue Sep 26 01:06:48 2017 -0600

M	packages/amesos2/CMakeLists.txt
M	packages/shylu/cmake/Dependencies.cmake
```

Closing this as complete!

@kyungjoo-kim thanks for fixing this so quickly and for using the checkin-test-sems.sh script for testing and pushing the change.
